xdem/vcrs.py

```diff
@@ -0,0 +1,256 @@
+"""Routines for vertical CRS transformation (fully based on pyproj)."""
+from __future__ import annotations
+
+import os
+import pathlib
+import warnings
+from typing import Literal, TypedDict
+
+import pyproj
+from pyproj import CRS, Transformer
+from pyproj.crs import BoundCRS, CompoundCRS, GeographicCRS, VerticalCRS
+from pyproj.crs.coordinate_system import Ellipsoidal3DCS
+from pyproj.crs.enums import Ellipsoidal3DCSAxis
+
+from xdem._typing import MArrayf, NDArrayf
+
+# Sources for defining vertical references:
+# AW3D30: https://www.eorc.jaxa.jp/ALOS/en/aw3d30/aw3d30v11_format_e.pdf
+# SRTMGL1: https://lpdaac.usgs.gov/documents/179/SRTM_User_Guide_V3.pdf
+# SRTMv4.1: http://www.cgiar-csi.org/data/srtm-90m-digital-elevation-database-v4-1
+# ASTGTM2/ASTGTM3: https://lpdaac.usgs.gov/documents/434/ASTGTM_User_Guide_V3.pdf
+# NASADEM: https://lpdaac.usgs.gov/documents/592/NASADEM_User_Guide_V1.pdf, HGTS is ellipsoid, HGT is EGM96 geoid !!
+# ArcticDEM (mosaic and strips): https://www.pgc.umn.edu/data/arcticdem/
+# REMA (mosaic and strips): https://www.pgc.umn.edu/data/rema/
+# TanDEM-X 90m global: https://geoservice.dlr.de/web/dataguide/tdm90/
+# COPERNICUS DEM: https://spacedata.copernicus.eu/web/cscda/dataset-details?articleId=394198
+vcrs_dem_products = {
+    "ArcticDEM": "Ellipsoid",
+    "REMA": "Ellipsoid",
+    "TDM1": "Ellipsoid",
+    "NASADEM-HGTS": "Ellipsoid",
+    "AW3D30": "EGM96",
+    "SRTMv4.1": "EGM96",
+    "ASTGTM2": "EGM96",
+    "ASTGTM3": "EGM96",
+    "NASADEM-HGT": "EGM96",
+    "COPDEM": "EGM08",
+}
+
+
+def _parse_vcrs_name_from_product(product: str) -> str | None:
+    """
+    Parse vertical CRS name from DEM product name.
+
+    :param product: Product name (typically from satimg.parse_metadata_from_fn).
+
+    :return: vcrs_name: Vertical CRS name.
+    """
+
+    if product in vcrs_dem_products.keys():
+        vcrs_name = vcrs_dem_products[product]
+    else:
+        vcrs_name = None
+
+    return vcrs_name
+
+
+def _build_ccrs_from_crs_and_vcrs(crs: CRS, vcrs: VerticalCRS | Literal["Ellipsoid"]) -> CompoundCRS | CRS:
+    """
+    Build a compound CRS from a horizontal CRS and a vertical CRS.
+
+    :param crs: Horizontal CRS.
+    :param vcrs: Vertical CRS.
+
+    :return: Compound CRS (horizontal + vertical).
+    """
+
+    # If a vertical CRS was passed, build a compound CRS with horizontal + vertical
+    # This requires transforming the horizontal CRS to 2D in case it was 3D
+    # Using CRS() because rasterio.CRS does not allow to call .name otherwise...
+    if isinstance(vcrs, CRS):
+        ccrs = CompoundCRS(
+            name="Horizontal: " + CRS(crs).name + "; Vertical: " + vcrs.name,
+            components=[CRS(crs).to_2d(), vcrs],
+        )
+    # Else if "Ellipsoid" was passed, there is no vertical reference
+    # We still have to return the CRS in 3D
+    else:
```

Comment:
This will run if `vcrs` is a sausage too. I'd prefer:
```python
elif vcrs == "Ellipsoid":
   ccrs = CRS(crs).to_3d()
else:
  raise ValueError(f"Invalid vcrs given. Must be a VerticalCRS or the literal string 'Ellipsoid'")

```

Reply:
I saw the tests now. I guess `str(vcrs).lower() == "ellipsoid"` is required to make the tests pass.

Reply:
Done! :D